File: leetcode/00286_walls-and-gates/286-walls-and-gates.cpp
class Solution {
public:
    void wallsAndGates(vector<vector<int>>& rooms) {
        // constants
        int EMPTY = pow(2, 31) - 1;
        int WALL = -1;
        int GATE = 0;

        int R = rooms.size(), C = rooms[0].size();
        queue<vector<int>> q;

        // gather gates
        for (int r = 0; r < R; r++) {
            for (int c = 0; c < C; c++) {
                if (rooms[r][c] == GATE) {
                    q.push(vector<int>{r, c, 0});
                }
            }
        }

        vector<pair<int, int>> offsets = {
            {1, 0},
            {-1, 0},
            {0, 1},
            {0, -1},
        };

        // bfs
        while (!q.empty()) {
            vector<int> node = q.front(); q.pop();
            int r = node[0];
            int c = node[1];
            int step = node[2];

            for (auto offset: offsets) {
                int _r = r + offset.first;
                int _c = c + offset.second;

                if (_r < 0 || _r >= R || _c < 0 || _c >= C || rooms[_r][_c] != EMPTY) {
                    continue;
                }

                rooms[_r][_c] = step + 1;
                q.push({_r, _c, step + 1});
            }            
        }
    }
};

File: leetcode/00286_walls-and-gates/286-walls-and-gates.py
class Solution:
    EMPTY = 2**31 - 1
    WALL = -1
    GATE = 0

    # time O(MN), space O(MN)
    def wallsAndGates(self, rooms: List[List[int]]) -> None:
        """
        Do not return anything, modify rooms in-place instead.
        """
        R, C = len(rooms), len(rooms[0])

        offsets = [
            (1, 0),
            (-1, 0),
            (0, 1),
            (0, -1),
        ]

        nodes = deque()
        for r in range(R):
            for c in range(C):
                if rooms[r][c] == self.GATE:
                    nodes.append((r, c, 0))

        # begin from each gates, run BFS
        while nodes:
            r, c, step_count = nodes.popleft()

            for r_offset, c_offset in offsets:
                _r = r + r_offset
                _c = c + c_offset

                if (
                    not (0 <= _r < R)
                    or not (0 <= _c < C)
                    or rooms[_r][_c] != self.EMPTY
                ):
                    continue

                rooms[_r][_c] = step_count + 1
                nodes.append((_r, _c, step_count + 1))
